File: leetcode_py/4Sum.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[][]}
    def fourSum(self, nums, target):
        nums.sort()
        res = set([])
        for i in range(len(nums)-3):
            for j in range(i+1, len(nums)-2):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue
                dict = {}
                for k in range(j+1, len(nums)):
                    if nums[k] in dict:
                        res.add((nums[i], nums[j], nums[dict[nums[k]]], nums[k]))
                    else:
                        dict[target-nums[k]-nums[i]-nums[j]] = k
        return map(list, res)


class Solution:
    def twosum(self,nums,target,res,idx1,idx2):
        dicts={}
        for i in range(idx2+1,len(nums)):
            if nums[i] in dicts:
                res.add((nums[idx1],nums[idx2],nums[dicts[nums[i]]],nums[i]))
            else:
                dicts[target-nums[i]]=i


    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[][]}
    def fourSum(self, nums, target):
        nums.sort()
        res=set([])
        for i in range(len(nums)-3):
            for j in range(i+1,len(nums)-2):
                self.twosum(nums,target-nums[i]-nums[j],res,i,j)
                # return list(res)
        r=[]
        for i in res:
            r.append(list(i))
        return r

File: leetcode_java/4sum/4Sum.java
public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> result = new  ArrayList<List<Integer>>();
        if(nums == null || nums.length <4){
            return result;
        }
        List<Integer> temp = new ArrayList<Integer>();
        int i , j , k , m;
        int len = nums.length;
        int sum;
        Arrays.sort(nums);
        for(i = 0 ; i < len - 3 ;i++){
           if((i != 0) && (nums[i] == nums[i - 1])){
                continue;
            }
            for(j = i + 1 ; j < len - 2 ;j++){
                if((j != i+1) &&nums[j] == nums[j - 1]){
                    continue;
                }
                k = j + 1;
                m = len - 1;
                while(k < m){
                    sum = nums[i] + nums[j] + nums[k] + nums[m];
                    if(sum == target){
                        temp = new ArrayList<Integer>();
                        temp.add(nums[i]);
                        temp.add(nums[j]);
                        temp.add(nums[k]);
                        temp.add(nums[m]);
                        result.add(temp);
                        k++;
                        m--;
                        while(k < m  && nums[m] == nums[m + 1]){
                        m--;
                        }
                     //System.out.println(m);
                        while(k < m && nums[k] == nums[k - 1]){
                             k++;
                         }
                         
                    }else if(sum < target){
                        k++;
                    }
                    else if(sum > target){
                        m--;
                    }
                    
                }
            
            }
        }
        return result;
    }
}